flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/type/TypeConverters.java

```diff
@@ -0,0 +1,108 @@
+import org.apache.flink.api.java.typeutils.PojoTypeInfo;
+import org.apache.flink.api.java.typeutils.RowTypeInfo;
+import org.apache.flink.api.java.typeutils.TupleTypeInfo;
+
+import com.google.common.collect.ImmutableBiMap;
+
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+/**
+ * Converters of {@link InternalType} and {@link TypeInformation}.
+ */
+public class TypeConverters {
+
+	public static final ImmutableBiMap<InternalType, TypeInformation> INTERNAL_TYPE_TO_EXTERNAL_TYPE_INFO =
+			new ImmutableBiMap.Builder<InternalType, TypeInformation>()
+					.put(InternalTypes.STRING, BasicTypeInfo.STRING_TYPE_INFO)
+					.put(InternalTypes.BOOLEAN, BasicTypeInfo.BOOLEAN_TYPE_INFO)
+					.put(InternalTypes.DOUBLE, BasicTypeInfo.DOUBLE_TYPE_INFO)
+					.put(InternalTypes.FLOAT, BasicTypeInfo.FLOAT_TYPE_INFO)
+					.put(InternalTypes.BYTE, BasicTypeInfo.BYTE_TYPE_INFO)
+					.put(InternalTypes.INT, BasicTypeInfo.INT_TYPE_INFO)
+					.put(InternalTypes.LONG, BasicTypeInfo.LONG_TYPE_INFO)
+					.put(InternalTypes.SHORT, BasicTypeInfo.SHORT_TYPE_INFO)
+					.put(InternalTypes.CHAR, BasicTypeInfo.CHAR_TYPE_INFO)
+					.put(InternalTypes.BYTE_ARRAY, PrimitiveArrayTypeInfo.BYTE_PRIMITIVE_ARRAY_TYPE_INFO)
+					.put(InternalTypes.DATE, SqlTimeTypeInfo.DATE)
+					.put(InternalTypes.TIMESTAMP, SqlTimeTypeInfo.TIMESTAMP)
+					.put(InternalTypes.TIME, SqlTimeTypeInfo.TIME)
+					.put(InternalTypes.SYSTEM_DEFAULT_DECIMAL, BasicTypeInfo.BIG_DEC_TYPE_INFO)
+					.build();
+
+	/**
+	 * Create a {@link InternalType} from a {@link TypeInformation}.
+	 *
+	 * <p>Note: Information may be lost. For example, after Pojo is converted to InternalType,
+	 * we no longer know that it is a Pojo and only think it is a Row.
+	 *
+	 * <p>Eg:
+	 * {@link BasicTypeInfo#STRING_TYPE_INFO} => {@link InternalTypes#STRING}.
+	 * {@link BasicTypeInfo#BIG_DEC_TYPE_INFO} => {@link DecimalType}.
+	 * {@link RowTypeInfo} => {@link RowType}.
+	 * {@link PojoTypeInfo} (CompositeType) => {@link RowType}.
+	 * {@link TupleTypeInfo} (CompositeType) => {@link RowType}.
+	 */
+	public static InternalType createInternalTypeFromTypeInfo(TypeInformation typeInfo) {
+
+		InternalType type = INTERNAL_TYPE_TO_EXTERNAL_TYPE_INFO.inverse().get(typeInfo);
+		if (type != null) {
+			return type;
+		}
+
+		if (typeInfo instanceof CompositeType) {
+			CompositeType compositeType = (CompositeType) typeInfo;
+			return InternalTypes.createRowType(
+					Stream.iterate(0, x -> x + 1).limit(compositeType.getArity())
+							.map((Function<Integer, TypeInformation>) compositeType::getTypeAt)
+							.map(TypeConverters::createInternalTypeFromTypeInfo)
+							.toArray(InternalType[]::new),
+					compositeType.getFieldNames()
+			);
+		} else if (typeInfo instanceof PrimitiveArrayTypeInfo) {
+			PrimitiveArrayTypeInfo arrayType = (PrimitiveArrayTypeInfo) typeInfo;
+			return InternalTypes.createArrayType(createInternalTypeFromTypeInfo(arrayType.getComponentType()));
```

Comment:
I think we should distinguish primitive array type and generic array type which will be used in code generation.

Actually, the primitive array is an ArrayType with element type is not null. So what do you think about to introduce a `isNullable` method to InternalType?  Which is closer to Calcite `RelDataType` and also helpful for code generation.

Reply:
Yes, We can do some optimization when we know there is no null.

Reply:
If we want to add a `isNullable`,  maybe it's nice to have a `sameType(InternalType)` method to check equals with ignoring nullable.

Reply:
I think add it when we need it. Now just go directly to getElementType.equals.